### PR TITLE
pretty printing: improved `Trace`

### LINKED ISSUE #9044: pretty printing: `Trace` could be improved (and LaTeX)
````
In [69]: X = ImmutableMatrix([[1, 2], [3, 4]])

In [70]: pprint(X)
⎡1  2⎤
⎢    ⎥
⎣3  4⎦

In [71]: w = Trace(X) + Trace(2*X)

In [72]: pprint(w)
Trace(Matrix([ + Trace(Matrix([
   [1, 2],          [2, 4],    
  [3, 4]]))        [6, 8]]))   
````

I suppose `pprint(Trace(X))` should be:
````
  ⎛⎡1  2⎤⎞ 
tr⎜⎢    ⎥⎟ 
  ⎝⎣3  4⎦⎠ 
````

### PR DESCRIPTION
**Pull request by swachhandl**
Attempts to fix #9044 
Defined a function _print_Trace in sympy.printing.pretty.pretty.PrettyPrinter
so that this function is used to print the Trace rather than the
emptyPrinter.

Example:

Before:

```
In [70]: X = ImmutableMatrix([[1, 2], [3, 4]])

In [71]: w = Trace(X) + Trace(2*X)

In [72]: pprint(w)
Trace(Matrix([ + Trace(Matrix([
   [1, 2],          [2, 4],
  [3, 4]]))        [6, 8]]))
```

After:

```
In [70]: X = ImmutableMatrix([[1, 2], [3, 4]])

In [71]: w = Trace(X) + Trace(2*X)

In [72]: pprint(w)
  ⎛⎡1  2⎤⎞ +   ⎛⎡2  4⎤⎞
tr⎜⎢    ⎥⎟   tr⎜⎢    ⎥⎟
  ⎝⎣3  4⎦⎠     ⎝⎣6  8⎦⎠
```

The position of the operators ('+' in the above example) in between the Traces are off position.
Still need to figure out how to fix that.

### CODE REVIEW

**Comment by aktech** [comment]
Please add relevant tests as well.

File: sympy/printing/pretty/pretty.py [hunk 1]
```diff
@@ -639,6 +639,14 @@ def _print_MatrixBase(self, e):
         D = self._print_matrix_contents(e)
         D = prettyForm(*D.parens('[', ']'))
         return D
+
+    def _print_Trace(self, e):
+        mat = e._args[0]
```
Comment (cbm755) [inline sympy/printing/pretty/pretty.py]: `e.arg`?

File: sympy/printing/pretty/pretty.py [hunk 2]
```diff
@@ -639,6 +639,14 @@ def _print_MatrixBase(self, e):
...
+        D = self._print_MatrixBase(mat)
```
Comment (cbm755) [inline sympy/printing/pretty/pretty.py]: would `._print(e.arg)` work?

File: sympy/printing/pretty/pretty.py [hunk 3]
```diff
@@ -639,6 +639,14 @@ def _print_MatrixBase(self, e):
...
+        D = prettyForm(*D.parens('(', ')'))
+        D = prettyForm(*D.left('\n'*(D.height()//2) + 'tr'))
+        return D
+
```
Comment (cbm755) [inline sympy/printing/pretty/pretty.py]: two blank lines before and after function def

File: sympy/printing/pretty/pretty.py (hunk 1 above)
Comment (swachhandl) [inline sympy/printing/pretty/pretty.py]: Yes. That would be better.

**Comment by cbm755** [comment]
Instead of dropping "tr" down, I think you need to lift the argument above the baseline.  Probably looking at how `pprint(sin(x/2) + 7)` works would be helpful (note I haven't read the code yet, so its just a suggestion).

**Comment by gxyd** [comment]
@swachhandl  can you please add test cases for changes introduced?

**Comment by cbm755** [comment]
I think other changes were merged to fix this.  Please do reopen if I'm wrong about that.

**Comment by gxyd** [comment]
I felt so from the conversation. But seeing the dates the PR was introduced and you commented in a progressive way, so I didn't close the PR.

I'll look at the code then.

**Comment by cbm755** [comment]
IIRC, this PR pre-dates the one that was eventually merged.

**Comment by gxyd** [comment]
Aah! its year 2015 not 2016. Yes, you are right in closing this. Thanks.

**Pull request closed without merging**